src/Compilers/CSharp/Portable/Symbols/Synthesized/GeneratedNameParser.cs

```diff
@@ -172,34 +173,56 @@ internal static bool TryParseAnonymousTypeParameterName(string typeParameterName
+        private const int sha256LengthHexChars = sha256LengthBytes * 2;
+
         // A full metadata name for a generic file-local type looks like:
         // <ContainingFile>FN__ClassName`A
-        // where 'N' is the syntax tree ordinal, 'A' is the arity,
+        // where 'N' is the SHA256 checksum of the original file path, 'A' is the arity,
         // and 'ClassName' is the source name of the type.
         //
         // The "unmangled" name of a generic file-local type looks like:
         // <ContainingFile>FN__ClassName
-        private static readonly Regex s_fileTypeOrdinalPattern = new Regex(@"<([a-zA-Z_0-9]*)>F(\d)+__", RegexOptions.Compiled);
+        private static readonly Regex s_fileTypeOrdinalPattern = new Regex($@"<([a-zA-Z_0-9]*)>F([0-9A-F]{{{sha256LengthHexChars}}})__", RegexOptions.Compiled);
 
         /// <remarks>
         /// This method will work with either unmangled or mangled type names as input, but it does not remove any arity suffix if present.
         /// </remarks>
-        internal static bool TryParseFileTypeName(string generatedName, [NotNullWhen(true)] out string? displayFileName, out int ordinal, [NotNullWhen(true)] out string? originalTypeName)
+        internal static bool TryParseFileTypeName(string generatedName, [NotNullWhen(true)] out string? displayFileName, [NotNullWhen(true)] out byte[]? checksum, [NotNullWhen(true)] out string? originalTypeName)
         {
-            if (s_fileTypeOrdinalPattern.Match(generatedName) is Match { Success: true, Groups: var groups, Index: var index, Length: var length }
-                && int.TryParse(groups[2].Value, out ordinal))
+            if (s_fileTypeOrdinalPattern.Match(generatedName) is Match { Success: true, Groups: var groups, Index: var index, Length: var length })
             {
                 displayFileName = groups[1].Value;
 
+                var checksumString = groups[2].Value;
+                var builder = new byte[sha256LengthBytes];
+                for (var i = 0; i < sha256LengthBytes; i++)
+                {
+                    builder[i] = (byte)((hexCharToByte(checksumString[i * 2]) << 4) | hexCharToByte(checksumString[i * 2 + 1]));
+                }
+                checksum = builder;
+
                 var prefixEndsAt = index + length;
                 originalTypeName = generatedName.Substring(prefixEndsAt);
                 return true;
             }
 
-            ordinal = -1;
+            checksum = null;
             displayFileName = null;
             originalTypeName = null;
             return false;
+
+            static byte hexCharToByte(char c)
+            {
+                return c switch
+                {
+                    >= '0' and <= '9' => (byte)(c - '0'),
+                    >= 'A' and <= 'F' => (byte)(10 + c - 'A'),
+                    _ => @throw(c)
+                };
+
+                static byte @throw(char c) => throw ExceptionUtilities.UnexpectedValue(c);
```

Comment:
>throw ExceptionUtilities.UnexpectedValue(c) [](http://example.com/codeflow?start=46&length=43)

Minor: Consider inlining if there's only one caller.

Reply:
I was doing this to try and make it so `hexCharToByte` is inlinable.